core/nodejsActionBase/src/service.js

```diff
@@ -44,112 +65,96 @@ function NodeActionService(config, logger) {
+                    return errorMessage(502, "The action did not return a dictionary.");
                 } else {
-                    res.status(500).json({ error: 'An error has occurred: ' + errMsg });
+                    return responseMessage(200, result);
                 }
+            }).catch(function (error) {
+                setStatus(Status.ready);
+
+                return Promise.reject(errorMessage(500, "An error has occurred: " + error));
             });
-        } else if (status === Status.starting || status === Status.running) {
-            logger.info('[runCode]', 're-scheduling runCode due to status',  status);
-            setTimeout(function() { runCode(req, res); }, 1); // Node can achieve 1 ms
         } else {
             logger.info('[runCode]', 'cannot schedule runCode due to status', status);
-            res.status(500).json({});
-        }
-    }
-
-    function shutdown(res, code, msg) {
-        status = Status.stopped
-        logger.info('[setStatus]', 'stopping now');
-        res.status(code).json(msg);
-        server.close();
-    }
-
-    function setStatus(newStatus) {
-        if (status !== Status.stopped) {
-            status = newStatus;
+            return Promise.reject(errorMessage(500, "Internal system error: container not ready."));
         }
     }
 
     function doInit(message) {
         var context = newWhiskContext(config, logger);
-        userScript = new NodeActionRunner(message, context);
-        if (typeof userScript.userScriptMain === 'function') {
+
+        userCodeRunner = new NodeActionRunner(context);
+
+        return userCodeRunner.init(message).then(function (result) {
             setStatus(Status.ready);
             return true;
-        } else {
+        }).catch(function (error) {
             writeMarkers();
             setStatus(Status.stopped);
-            return false;
-        }
+            return Promise.reject(error);
+        });
     }
 
-    function doRun(req, next) {
-        try {
-            var ids = (req.body || {}).meta;
-            var args = (req.body || {}).value;
-            var authKey = (req.body || {}).authKey;
-            userScript.whisk.setAuthKey(authKey)
-            userScript.run(args, function(response) {
-                writeMarkers();
-                next(response, undefined);
-            });
-        } catch (e) {
-            console.error(e);
+    function doRun(req) {
+        var ids = (req.body || {}).meta;
+        var args = (req.body || {}).value;
+        var authKey = (req.body || {}).authKey;
+        userCodeRunner.whisk.setAuthKey(authKey)
+
+        return userCodeRunner.run(args).then(function(response) {
             writeMarkers();
-            next(false, e);
-        }
+            return response;
+        }).catch(function (error) {
+            console.error(error);
+            writeMarkers();
+            reject(e);
```

Comment:
There is no `reject` function available here. Seems like perhaps a leftover from your previous implementation? Probably meant to be `return Promise.reject(e)`